### ISSUE TITLE
Questions about joint score

### ISSUE THREAD

**Issue by MichaelH717** [OpenDriveLab/BeTop#4]
hello, thanks for your release code, I was confused about joint score function, is it true for just select M largest joint distributions for all NI interacted agents?

![Image](https://github.com/user-attachments/assets/8d2dd00c-381d-48e2-8159-d190c167d3b5)

**Comment by georgeliu233** [comment]
Yes, your understanding is correct. In our project, it refers to a faster and tensor-wise implementation, to replace the standard for-loop.

**Comment by MichaelH717** [comment]
> Yes, your understanding is correct. In our project, it refers to a faster and tensor-wise implementation, to replace the standard for-loop.

thanks a lot, I write a demo code for this implemention, is it right? can you show the pseudo code? @georgeliu233 
`import numpy as np

def compute_joint_distribution(p_list):
    N = len(p_list)           # 障碍物数 N_I
    M = p_list[0].shape[0]    # 每个障碍物的模态数
    shape = (M,) * N         # 张量形状 (M, M, ..., M)
    PJ = np.ones(shape)       # 初始化所有组合的概率为 1
    
    # 对每个障碍物，将其模态概率通过广播乘到 PJ 上
    for n, p in enumerate(p_list):
        # 构造形状为 (1,1,...,M,...,1) 的张量，使得 p 在第 n 维有 M 个元素，其它维度为 1
        shape_n = [1] * N
        shape_n[n] = M
        p_n = p.reshape(shape_n)
        PJ = PJ * p_n  # 广播乘法，相当于对每个组合累乘对应障碍物的概率
        
    return PJ

def viterbi_max_product(p_list):
    N = len(p_list)         # 障碍物数
    M = p_list[0].shape[0]  # 每个障碍物模态数
    shape = (M,) * N        # 初始张量形状 (M, M, ..., M)
    PJ = np.ones(shape)     # 初始化张量，所有元素为 1
    backpointers = []       # 用于存储每一步的 argmax（回溯指针）
   
    for n, p in enumerate(p_list):
        # 当前 PJ 的维度数（初始为 N，之后依次递减）
        current_ndim = PJ.ndim 
        # 将 p 重塑为形状 (M, 1, 1, ..., 1) 以便在第 0 维上进行广播
        p_n = p.reshape((M,) + (1,)*(current_ndim - 1))
        # 乘法更新
        PJ = PJ * p_n
        # 在第 0 维上取最大值，同时记录 argmax（此操作消除第 0 维）
        bp = np.argmax(PJ, axis=0)
        backpointers.append(bp)
        PJ = np.max(PJ, axis=0)
    
    # 此时 PJ 已经退化为标量，即最佳联合得分
    best_score = PJ.item()
    best_sequence = [int(np.argmax(p)) for p in p_list]
    
    return best_score, best_sequence

# === 测试示例 ===
if __name__ == "__main__":
    # 假设有 4 个障碍物，每个障碍物有 6 个模态
    N = 4
    M = 6
    p_list = []
    for i in range(N):
        # 随机生成概率向量，并归一化
        p = np.random.rand(M)
        p = p / np.sum(p)
        p_list.append(p)
    
    # 1. 计算联合概率分布张量
    PJ = compute_joint_distribution(p_list)
    print("【联合概率分布】")
    print("PJ 的形状:", PJ.shape)
    print("PJ 的前 10 个数值:", PJ.flatten()[:10])
    
    # 2. 利用迭代广播+max操作计算最大联合得分和最佳模态组合
    best_score, best_sequence = viterbi_max_product(p_list)
    print("\n【最大乘积(max-product)计算】")
    print("最大联合得分:", best_score)
    print("最佳模态组合:", best_sequence)`

**Comment by georgeliu233** [comment]
Hi, I think the logic is correct. Here is the part of my implementation:

```
def joint_combination(pred, score, mask, top_n=6):
    '''
    conducting full combinations across all modal 
    and keep the top-k joint modes
    pred: [b, n, m, t, d]; score: [b, n, m]; mask: [b, n]
    '''
    b, n, m, t, d = pred.shape
    m_inds = torch.arange(m).to(score.device)
    mask = mask.float()[:, :, None]
    # the masked score for certain agent will be 1 for its all modes
    score = score * mask 
    score = score + (1 - mask)
    # [b, m]
    m_inds = m_inds[None, :].repeat(b, 1)
    comb_inds_shape = (b,) + (n,) + (m,) * n
    comb_score_shape =  (b,) + (m,) * n 
    vanilla_inds = torch.zeros(comb_inds_shape).to(score.device)
    vanilla_scores = torch.ones(comb_score_shape).to(score.device)

    for i in range(n):
        insert_score_shape = (b,) + (1,)*i + (m,) + (1,)*(n-i-1)
        vanilla_scores = vanilla_scores * score[:, i].reshape(insert_score_shape)
        vanilla_inds[:, i] = vanilla_inds[:, i] + m_inds.reshape(insert_score_shape)

    vanilla_scores, vanilla_inds = torch.flatten(vanilla_scores, start_dim=1), torch.flatten(vanilla_inds, start_dim=2)
    top_score, top_score_inds = vanilla_scores.topk(top_n, dim=-1)
    top_score_inds = top_score_inds[:, None, :].repeat(1, n, 1)
    #[b, n, top_n]
    top_n_inds = vanilla_inds[torch.arange(b)[:, None, None], torch.arange(n)[None, :, None], top_score_inds]
    top_n_inds = top_n_inds.long()
    ret_traj = pred[torch.arange(b)[:, None, None], torch.arange(n)[None, :, None], top_n_inds]
    return ret_traj, top_score
```
